File: Python/bubble_sort.py
from typing import List


def bubble_sort(array: List[int]) -> None:
    """Bubble Sort algorithm implementation in python

    Time complexity: O(n^2)
    Space complexity: O(1)

    Args:
        array (List[int]): Array of integers to be sorted
    """

    while True:

        swapped = False

        for i in range(len(array) - 1):
            if array[i] > array[i + 1]:

                # pythonic way to swap
                array[i], array[i + 1] = array[i + 1], array[i]
                swapped = True

        if not swapped:
            break


if __name__ == "__main__":
    testArray = list(
        map(
            int, input("Enter space separated array of integers to be sorted: ").split()
        )
    )
    bubble_sort(testArray)
    print("Sorted array =", testArray)

File: cpp_algorithms/bubble_sort.cpp
#include <iostream>
using namespace std;

/**
 * Bubble sort algorithm.
 * Time complexity: O(n^2),
 * Space complexity: O(1)
 *
 * @param arr[] array of integers to be sorted
 * @param size size of array
 * @return null, sorts array in place
 */
void bubbleSort(int arr[], int size)
{
    while (true)
    {

        bool swapped = false;

        for (int i = 0; i < size - 1; i++)
        {
            if (arr[i] > arr[i + 1])
            {

                swapped = true;

                int temp = arr[i];
                arr[i] = arr[i + 1];
                arr[i + 1] = temp;
            }
        }

        if (!swapped)
        {
            break;
        }
    }
}

int main()
{

    int arr[] = {34, 121, 454, 22, 214, 54};
    int n = sizeof(arr) / sizeof(arr[0]);
    bubbleSort(arr, n);

    cout << "Sorted array: ";

    for (int i = 0; i < n; i++)
    {
        cout << arr[i] << " ";
    }

    cout << endl;
    return 0;
}
